Janestreet

maurer opened the pull request:
###### Motivation for this change

OCaml has switched its primary macro system in OCaml 4.02 to PPX (from camlp4).
Jane Street, who produces one of the OCaml basis libraries, took this opportunity to change their build system, rip out all of their camlp4 stuff, and replace it with PPX.
New libraries and programs which depend on the new PPX style macros are being released, so we need a way to support them.
###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [x] Linux
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

To deal with older compilers and packages, I did a few things in addition to packaging the new code:
- Each package which has transitioned to PPX has been renamed to packagename_p4, allowing its last camlp4 release to be referenced explicitly
- Each package which transitioned to PPX had an if statement added which will diversion to packagename_p4 if the compiler is too old, so OCaml 4.0.0, 3.1, etc will still be able to access e.g. "sexplib" rather than "sexplib_p4"
- Existing packages which referenced a transitioning package were pointed at the _p4 version.

maurer committed:
js_build_tools: init at 113.33.06

maurer committed:
ppx_core: init at 113.33.03

maurer committed:
ppx_optcomp: init at 113.33.03

maurer committed:
ppx_driver: init at 113.33.03

maurer committed:
ppx_deriving: init at v3.3

maurer committed:
ppx_type_conv: init at 113.33.03

maurer committed:
ppx_compare: init at 113.33.03

maurer committed:
ppx_here: init at 113.33.03

maurer committed:
ppx_sexp_conv: init at 113.33.03

maurer committed:
ppx_assert: init at 113.33.03

maurer committed:
ppx_inline_test: init at 113.33.03

maurer committed:
ppx_bench: init at 113.33.03

maurer committed:
ppx_bin_prot: init at 113.33.03

maurer committed:
ppx_custom_printf: init at 113.33.03

maurer committed:
ppx_enumerate: init at 113.33.03

maurer committed:
ppx_fail: init at 113.33.03

maurer committed:
ppx_fields_conv: init at 113.33.03

maurer committed:
ppx_let: init at 113.33.03

maurer committed:
ppx_pipebang: init at 113.33.03

maurer committed:
ppx_sexp_message: init at 113.33.03

maurer committed:
ppx_sexp_value: init at 113.33.03

maurer committed:
ppx_typerep_conv: init at 113.33.03

maurer committed:
ppx_variants_conv: init at 113.33.03

maurer committed:
ppx_expect: init at 113.33.03

maurer committed:
ocaml-typerep: 112.24.00 -> 112.24.00/113.33.03

Split typerep into ppx supporting (modern) and p4 supporting (legacy)

maurer committed:
ppx_jane: init at 113.33.03

maurer committed:
fieldslib: 109.20.03 -> 109.20.03/113.33.03

PPX/P4 legacy split

maurer committed:
sexplib: 112.24.01 -> 112.24.01/113.33.03

PPX/P4 split

maurer committed:
variantslib: 109.15.03 -> 109.15.03/113.33.03

PPX/P4 split

maurer committed:
bin_prot: 112.24.00 -> 112.24.00/113.33.03

PPX/P4 split

maurer committed:
core_kernel: 112.24.00 -> 112.24.00/113.33.03

PPX/P4 split

maurer committed:
core: 112.24.01 -> 112.24.01/113.33.03

PPX/P4 split

maurer committed:
ocaml-re2: 112.06.00 -> 112.06.00/113.33.03

PPX/P4 split

maurer committed:
ocaml-textutils: 112.17.00 -> 112.17.00/113.33.03

PPX/P4 split

maurer committed:
core_extended: 112.24.00 -> 112.24.00/113.33.03

PPX/P4 split

maurer committed:
Legacy library + old compiler = normal name (OCaml)

For legacy camlp4 libraries, if we are using a pre-4.02 compiler, and
so don't have ppx access, allow access to the legacy library under the
normal name.

maurer committed:
Fix descriptions to be CONTRIBUTIONS.md compliant

maurer commented:
@vbgl Tagging you as another potential reviewer, feel free to ignore if you're not interested.

regnat commented on pkgs/development/ocaml-modules/janestreet/ppx-sexp-conv.nix:
```diff
@@ -0,0 +1,14 @@
+  propagatedBuildInputs = [ ppx_core ppx_tools ppx_type_conv sexplib];
+
+  meta = with stdenv.lib; {
+    description = "A ppx rewriter that defines an extension node whose value is its source position";
```
This description doesn't seem to correspond to the ppx

regnat commented on pkgs/development/ocaml-modules/janestreet/buildOcamlJane.nix:
```diff
@@ -0,0 +1,30 @@
+
+  installPhase = ''
+    opam-installer -i --prefix $prefix --libdir `ocamlfind printconf destdir` --stubsdir `ocamlfind printconf destdir`/${name} ${name}.install
+    if [ -d $out/lib/${name} ]; then if [ "$(ls -A $out/lib/${name})" ]; then mv $out/lib/${name}/* `ocamlfind printconf destdir`/${name}; fi; rmdir $out/lib/${name}; fi
```
This would gain in readability with a few more line-breaks

regnat commented:
Wow, that's a huge work !

I added a few minor annotations, otherwise it's a big :+1:

aske commented:
Right now I'm working on moving ocaml packages from all-packages.nix to make it look more like current haskell system; would be nice to see this PR merged first so I can deal with merge conflicts sooner ( the opposite way is fine too :) ).

maurer committed:
ppx_sexp_conv: Fix description to match package

maurer committed:
buildOcamlJane: Add line breaks

maurer committed:
async_kernel: 112.24.00 -> 112.24.00/113.33.03

PPX/P4 split

maurer committed:
async_unix: 112.24.00 -> 112.24.00/113.33.03

PPX/P4 split

maurer committed:
async_rpc_kernel: init at 113.33.03

maurer committed:
async_extra: 112.24.00 -> 112.24.00/113.33.03

PPX/P4 split

maurer committed:
async: 112.24.00 -> 112.24.00/113.33.03

PPX/P4 split

joachifm commented:
Is the darwin failure reported by travis legit and if so is it a regression?

vbgl commented:
I can reproduce the error on darwin: `ppx_core` won’t build with the following message.

```
+ xargs stat -f '%d:%i:%m' < .compiler-libs.common.files > .compiler-libs.common.stats
stat: cannot read file system information for '%d:%i:%m': No such file or directory
```

I have no idea where that comes from.

maurer commented:
I don't have easy access to a darwin machine, but `ppx_core` as a piece of software does work on OSX outside of this packagaing.

@joachifm It's not a regression since `ppx_core` is not in nixpkgs yet, but should probably be fixed before merging since I know the packages I'm adding/updating work on OSX when installed via OPAM.

maurer commented:
https://github.com/janestreet/js-build-tools/blob/7ff35bb6451f241bb1a643f73f73d9d55fd36d94/ocamlbuild_goodies/jane_street_ocamlbuild_goodies.ml#L66-L74

It looks like they are expecting some BSD-like version of `stat` for which `-f` means something else, that is normally present on OSX. With Darwin + nix, it may be getting a more Linuxish `stat`.

I can't test on Darwin easily, but does someone who can want to test removing `| Darwin` from that top match and see if it fixes things? @vbgl

vbgl commented:
Good catch. The following patch fixes the build.

```
--- a/ocamlbuild_goodies/jane_street_ocamlbuild_goodies.ml
+++ b/ocamlbuild_goodies/jane_street_ocamlbuild_goodies.ml
@@ -65,7 +65,7 @@ let track_external_deps = function

     let stat, md5sum =
       match run_and_read "uname" |> String.trim with
-      | "Darwin" ->
+      | "FreeBSD" | "NetBSD" ->
         (S [A "stat"; A "-f"; A "%d:%i:%m"],
          A "md5")
       | _ ->
```

maurer committed:
js-build-tools: Fix darwin build

On darwin, js-build-tools expects an OSX style userland and stat binary.
However, under nix we use the nix version of stat, which speaks linux-style
flags.

This patch removes js-build-tools special casing so that it speaks to stat
in linux style.

maurer commented:
I've applied the patch to fix the darwin build. Anything else to be done?
